cluster-autoscaler/core/loop_state.go

```diff
@@ -0,0 +1,448 @@
+		return ls.onError(errors.ApiCallError, err)
+	}
+	metrics.UpdateUnschedulablePodsCount(len(allUnschedulablePods))
+
+	ls.allScheduled, err = ls.autoscaler.ScheduledPodLister().List()
+	if err != nil {
+		glog.Errorf("Failed to list scheduled pods: %v", err)
+		return ls.onError(errors.ApiCallError, err)
+	}
+
+	allUnschedulablePods, ls.allScheduled, err = ls.autoscaler.podListProcessor.Process(
+		ls.context, allUnschedulablePods, ls.allScheduled, ls.allNodes)
+	if err != nil {
+		glog.Errorf("Failed to process pod list: %v", err)
+		return ls.onError(errors.InternalError, err)
+	}
+
+	ConfigurePredicateCheckerForLoop(allUnschedulablePods, ls.allScheduled, ls.autoscaler.PredicateChecker)
+
+	// We need to check whether pods marked as unschedulable are actually unschedulable.
+	// It's likely we added a new node and the scheduler just haven't managed to put the
+	// pod on in yet. In this situation we don't want to trigger another scale-up.
+	//
+	// It's also important to prevent uncontrollable cluster growth if CA's simulated
+	// scheduler differs in opinion with real scheduler. Example of such situation:
+	// - CA and Scheduler has slightly different configuration
+	// - Scheduler can't schedule a pod and marks it as unschedulable
+	// - CA added a node which should help the pod
+	// - Scheduler doesn't schedule the pod on the new node
+	//   because according to it logic it doesn't fit there
+	// - CA see the pod is still unschedulable, so it adds another node to help it
+	//
+	// With the check enabled the last point won't happen because CA will ignore a pod
+	// which is supposed to schedule on an existing node.
+	ls.scaleDownForbidden = false
+
+	unschedulablePodsWithoutTPUs := tpu.ClearTPURequests(allUnschedulablePods)
+
+	// Some unschedulable pods can be waiting for lower priority pods preemption so they have nominated node to run.
+	// Such pods don't require scale up but should be considered during scale down.
+	var unschedulablePods []*apiv1.Pod
+	unschedulablePods, ls.unschedulableWaitingForLowerPriorityPreemption = FilterOutExpendableAndSplit(
+		unschedulablePodsWithoutTPUs, ls.autoscaler.ExpendablePodsPriorityCutoff)
+
+	glog.V(4).Infof("Filtering out schedulables")
+	filterOutSchedulableStart := time.Now()
+	ls.unschedulablePodsToHelp = FilterOutSchedulable(unschedulablePods, ls.readyNodes, ls.allScheduled,
+		ls.unschedulableWaitingForLowerPriorityPreemption, ls.autoscaler.PredicateChecker, ls.autoscaler.ExpendablePodsPriorityCutoff)
+	metrics.UpdateDurationFromStart(metrics.FilterOutSchedulable, filterOutSchedulableStart)
+
+	if len(ls.unschedulablePodsToHelp) != len(unschedulablePods) {
+		glog.V(2).Info("Schedulable pods present")
+		ls.scaleDownForbidden = true
+	} else {
+		glog.V(4).Info("No schedulable pods")
+	}
+
+	return ls.onSuccess()
+}
+
+// See LoopStateHandler interface
+func (ls *loopState) ScaleUp() (ExecutionResult, errors.AutoscalerError) {
+	if len(ls.unschedulablePodsToHelp) == 0 {
```

Comment:
Instead of multiple else ifs and success path code contained in final else, can we have a series of ifs with early return? Like this:

if () {
 return
}
if () {
 return
}
// success path

Reply:
Yes, we can, committed.

I've copied the code structure inside the new methods so it looks more similar to the old one - but we can improve on it, of course.